### ISSUE TITLE
[Feature Request] bspc exec - Have programs appear in the desktops they are executed in

### ISSUE THREAD

**Issue by SeerLite** [baskerville/bspwm#1109]
Similar to [i3's `exec`](https://i3wm.org/docs/userguide.html#exec) without `--no-startup-id`
> The --no-startup-id parameter disables startup-notification support for this particular exec command. With startup-notification, i3 can make sure that a window appears on the workspace on which you used the exec command. Also, it will change the X11 cursor to watch (a clock) while the application is launching.

Is it possible to implement such a thing in bspwm? Consider the ability to be able to apply rules too.

Here's some concepts I came up with:
1. `bspc exec [--rules <rule>[,rule ...]] -- <program> [arguments ...]`
2. `bspc exec [--rule <rule> ...] -- <program> [arguments ...]`
And examples:
1. `bspc exec --rules desktop=.focused,state=floating,focus=off -- alacritty`
2. `bspc exec --rule desktop=.focused --rule state=floating --rule focus=off -- alacritty`

Note: Currently `DESKTOP_SEL` seem doesn't support plain `.focused` like `NODE_SEL`. I omitted `any` for readability.

**Comment by emanuele6** [comment]
> And no, using focused isn't the same as .focused. In DESKTOP_SEL, .focused actually refers to the focused monitor, not desktop. See the section that documents DESKTOP_SEL in bspwm(1) for information on that.

No, it selects the focused desktop: how can a desktop selector select a monitor?
Here is a screenshot of the documentation ![desktop_selector](https://user-images.githubusercontent.com/20175435/81249182-505e5700-901e-11ea-83f5-88becd3a47f9.png)

**Comment by Lythenas** [comment]
You could use one-shot rules for this. Although they are not perfect. Something like this might work:

`bspc rule --add "*" --one-shot desktop=focused state=floating focus=off && alacritty`

Note: The `*` is required (or any other `(<class_name>|*)[:(<instance_name>|*)]`).

However this only works if no other window is opened between exectuting the `bspc` command and the application opening its window. Also apps with multiple windows require more work. But you could refine this for some applications a bit so it will even work when another random window appears. This only works for apps that allow setting the class/instance name on start.

`bspc rule --add "my_new_window_123" --one-shot desktop=focused state=floating focus=off && alacritty --class "Alacritty,my_new_window_123"`

**Comment by emanuele6** [comment]
> However this only works if no other window is opened between exectuting the bspc command

It also gets weird when, in this case, `alacritty` fails to open and the one-shot rule stays.

**Comment by Lythenas** [comment]
You can remove the rule again with `bspc rule -r (<class_name>|*)[:(<instance_name>|*)]` using the same class/instance name you specified when creating it.

But I agree. there are a lot of corner cases where stuff can go wrong or get stuck.

I not 100% sure that bspwm should handle all this. Also maybe there is already a tool that does this. In my mind you would need the following from a `bspc exec` command:

1. execute the application and track all windows (or just the first) that are opened
2. apply the rules to those windows

It may be easier to do this inside bspwm but you may be able to do this in a separate tool (but I'm already struggling with the first part...).

**Comment by emanuele6** [comment]
Oh, I didn't realise you answered, I wouldn't have deleted that comment.  
Yes. I know I can remove the rule that way, but why would I want to do that? That's still weird, I wanted to use `bspc exec` to remove weirdness. What if the window takes a long time to open? I can't just wait some seconds and then remove the rule;  
Also what if I have a non-one-shot rule for that `class:instance`? It would get removed.

I deleted my comment because I figured it wasn't a very easy thing to implement (if not impossible).  
Let's say you have this (I actually use this):
```
bspc rule -a Urxvt:urxvt -o state=floating
urxvtc -e tmux
```
And you replace it with:
```
bspc exec --rule state=floating urxvtc
```

`urxvtc` will just send a message to `urxvtd` to tell it to spawn an `urxvt` window, it doesn't actually open the window from its `urxvtc` process, so how can `bspwm` know that that window was opened because of `urxvtc`?

I thought it was possible to implement because @SeerLite said..
> Similar to [i3's `exec`](https://i3wm.org/docs/userguide.html#exec)

..but then I thought about it, read the link and saw that `i3`'s `exec` doesn't actually allow you to set rules: it just executes a program (`i3` has that command because it apparently is configured with a configuration file with its own syntax, not with a script in any language like bspwm); and I realised it wasn't actually possible to implement relyably.

**Comment by Lythenas** [comment]
Yes I think there is no solution that works for everything. You probably need some ugly hacks to achieve this. Also there are a lot of programs that don't allow you to set the class/instance name on start.

I would use a randomly generated instance name or class name so I can be sure that I don't remove other one-shot rules. Also I would expect that `alacritty` (and also `urxvtc`) would return an error when they fail to open a window. So you could do something like this:
```
X="one_shot_$RANDOM"
bspc rule -a "$X" -o state=floating
alacritty --class "Alacritty,$X" || bspc rule -r "$X"
```

**Comment by emanuele6** [comment]
`urxvtc` (urxvt client) just sends a message to `urxvtd` (urxvt deamon) telling it to open a window AFAIK.

But still, even if it exited with success only after the `urxvt` window opened, that would be specific to `urxvtc`, `bspc exec` cannot know that.

**Comment by SeerLite** [comment]
First of all, sorry for the wall of text. I'm just really interested in this feature and wanted to respond to all the points you two made while I was asleep lol.

> You could use one-shot rules for this. Although they are not perfect.

Yeah, I know about them. I also recommend them a lot to people who come from other WMs that tend to modify the class (`--class=<class_name>`) of the applications they're launching just so the rules work out, when a simple oneshot rule is usually better.

Still, I think something like `bspc exec` would be a lot more reliable and feel less "hacky".

> You could also get the currently focused desktop via bspc query -D focused and use it in the desktop= part of the rule. But I think this command is bugged or I'm understanding something wrong. (because it returns all desktop ids)

I'm not sure if what I said about `DESKTOP_SEL` is correct though:

I originally thought `focused` was monitor-specific because of the `[MONITOR_SEL:]` that's right before it, but now I'm not sure anymore. `focused` as a descriptor is documented just a little bit below `DESKTOP_SEL` and it does indeed refer to the focused desktop. I think I just suck at reading docopt/the man page hah

Anyway, the correct syntax for your command is `bspc query -D -d focused`, or more abbreviated: `bspc query -D -d`.

> I not 100% sure that bspwm should handle all this. Also maybe there is already a tool that does this. In my mind you would need the following from a `bspc exec` command:
> 
> 1. execute the application and track all windows (or just the first) that are opened
> 2. apply the rules to those windows

A wrapper able to do this (not necessarily part of `bspc`) is exactly what I'm looking for. i3 does this by using startup notifications, a protocol that tells the WM when a program has launched.  
Disclaimer: I don't know much about this protocol. I'd like to investigate more about it though.

> I didn't realise you answered, I wouldn't have deleted that comment if I knew.

Hey, I don't think you should be deleting comments honestly. People who get their GitHub notifications through e-mail (like myself) are still gonna get the comments there. Also, your first reply here lead me to realize that I might be wrong about the `DESKTOP_SEL` thing.

> `urxvtc` will just send a message to `urxvtd` to tell it to spawn an `urxvt` window, it doesn't actually open the window from its `urxvtc` process, so how can `bspwm` know that that window was opened because of `urxvtc`?

Only programs that implement the startup notification protocol would be handled by `bspc exec`. i3 also has this problem (just tested it out). For example, Termite starts in the desktop it's executed in, while Alacritty does not. This is because Alacritty doesn't implement startup notifications (I wish they did though): [Alacritty issue about this I found just now](https://github.com/alacritty/alacritty/issues/2824)

> > Similar to [i3's `exec`](https://i3wm.org/docs/userguide.html#exec)
> 
> ..but then I thought about it, read the link and saw that `i3`'s `exec` doesn't actually allow you to set rules: it just executes a program (`i3` has that command because it apparently is configured with a configuration file with its own syntax: not with a script in any language like bspwm; so it needs a way to open external programs from the configuration file); and I finally realised it wasn't actually possible to implement relyably.

Oops, I should have clarified that i3 only implements workspace stuff. But I guessed it might be just as easy to implement rules, right? Ah, maybe this is indeed to difficult to implement.

Either way, i3 having its own config syntax shouldn't matter at all here. After all, you can still communicate with i3 through IPC (`i3-msg exec`) which I think is really similar to what `bspc` does to communicate with `bspwm`. Everything that'd matter here is that bspwm (or some external tool) keeps track of the programs that are launched and their startup notifications, in order to apply the right rules.

> Also I would expect that `alacritty` (and also `urxvtc`) would return an error when they fail to open a window.

Try any terminal with parameters like `-e false`. Most wont open a window because `false` just runs way too quickly for it to be worth it. And they'll return `0` too, since the program actually run (even if `false` itself returns `1`)

**Comment by emanuele6** [comment]
> After all, you can still communicate with i3 through IPC (i3-msg exec)

What I said is that if you wanted to start a program, for example a `alacritty`, from i3's configuration file, you could just write `alacritty` in one line, there needs to be an `i3` command to execute a process to do that: that's the main reason why they have an `exec` command, I think.

> Try any terminal with parameters like -e false. Most wont open a window because false just runs way too quickly for it to be worth it. And they'll return 0 too, since the program actually run (even if false itself returns 1)

You are missing the point: a window is still created even if it is not show because `false` exited to quickly. Try this:
* add a one-shot rule, `bspc rule -a '*:*' -o state=floating`;
* check that it's actually added, `bspc rule -l`;
* run `urxvtc -e false` or `alacritty -e false` or whatever;
* see if the rule is still there, `bspc rule -l`
That rule is gone, because it was used by the window.

**Comment by SeerLite** [comment]
Ah, I understand.
Also, you're right! Sorry again. I don't know why I didn't decide to test the `-e false` rules before posting.

**Comment by SeerLite** [comment]
So I looked a bit into the ["startup notification" specification](https://specifications.freedesktop.org/startup-notification-spec/startup-notification-latest.txt) and it looks like it only keeps track of desktops/workspaces so not all rules can be implemented this way. (unless there's a way to also get the window id out of it... hmm)

Still, maybe `bspc exec` could be used to launch programs on the current desktop, and bspwm could pay attention to the startup notification and place the started program on the right desktop (if it doesn't already)
